### PR TITLE
Introduce setup_default_system_test_config attribute macro

### LINKED ISSUE #3335: Simplify test setup
They are many duplicated setup code required for any new modules, mostly config `system::Trait`. There must be a way to avoid duplicate this code all over the places.
It also make any modification to `system::Trait` tedious as it have to change every single module test setup.

I am not sure what is the best way, but at least a macro can avoid most of the duplications.

https://github.com/paritytech/substrate/blob/9b3e9f32cd5243d2eeff3d5d8337033aa64c091b/srml/aura/src/mock.rs#L35-L63

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #3335.

This PR adds the ability to tag the `construct_runtime!` macro with an attribute called `setup_default_test_parameters`, which generates an implementation of the `frame_system::Config` trait on the runtime struct.

I will still need to work on the following task before I think it is complete:
- [x] Allow the attribute macro to accept attributes to configure the default associated types for the `frame_system::Config` implementation
- [ ] Documentation on what the defaults are and the usage

### CODE REVIEW

**Review by kianenigma — COMMENTED** [review]
I like the direction, but I think there'd be more revisions on this. 

important note: the naming now is not good: `setup_default_test_parameters` only sets up `system`. It should be called `with_default_test_system` or sth like that.  

minor note: My odyssey idea was that each pallet would itself declare what are its default values for its `trait Config` and you can only overwrite them in `construcy_runtime!` if you want.

**Comment by KiChjang** [comment]
I think it's good for now to have a solution where we do away with the most laborious process of setting up the test framework, namely configuring the `frame_system::Config` trait.

That said, the default value attribute sure sounds intriguing, but does require a fair bit more discussion around how exactly it should be implemented, as it does change the way in which developers use the `construct_runtime!` macro.

File: frame/support/procedural/src/test_setup.rs [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+		.as_ref()
+		.map(ToTokens::to_token_stream)
+		.unwrap_or(quote! {BlockHashCount});
+	let block_length =
+		attr.block_length.as_ref().map(ToTokens::to_token_stream).unwrap_or(quote! {()});
+	let block_number = attr
+		.block_number
+		.as_ref()
+		.map(ToTokens::to_token_stream)
+		.unwrap_or(quote! {u64});
```
Comment (bkchr) [inline frame/support/procedural/src/test_setup.rs:187]: Maybe we could directly write a `Default for TestSetupAttr` that does this.

File: frame/support/procedural/src/test_setup.rs [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+		Some(segment) if segment.ident == "construct_runtime" => {},
+		_ => {
+			let msg =
+				"Invalid test setup attribute macro: expected `construct_runtime!` macro call";
+			let span = mac.path.span();
+			return Err(syn::Error::new(span, msg))
+		},
+	}
+
+	let frame_system = generate_crate_access_2018("frame-system")?;
```
Comment (bkchr) [inline frame/support/procedural/src/test_setup.rs:151]: I'm no big fan of requiring random crates, that are not the ones that re-export the macro.

We should at least improve the error message and to say why this required etc.

File: frame/support/procedural/src/test_setup.rs [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+	let frame_support = generate_crate_access_2018("frame-support")?;
+	let sp_runtime = generate_crate_access_2018("sp-runtime")?;
```
Comment (bkchr) [inline frame/support/procedural/src/test_setup.rs:153]: This should already be re-exported by `frame-support`.

**Review by bkchr — COMMENTED** [review]
I think in general a dedicated macro call would be better. If you start changes types via the attributes of the macro, it could start very easily to look messy imo.

Maybe we could call it `construct_test_runtime` or whatever.

**Comment by apopiak** [comment]
> Maybe we could call it `construct_test_runtime` or whatever.

Maybe it could even just be part of `frame-system`?

**Comment by bkchr** [comment]
> > Maybe we could call it `construct_test_runtime` or whatever.
> 
> Maybe it could even just be part of `frame-system`?

I would say a new crate maybe :P

**Comment by KiChjang** [comment]
Huh, I just figured that maybe we can even modify the `construct_runtime!` macro to support some custom syntax to configure the test config, something like so:

```rust
construct_runtime!(
	pub enum Runtime where
		Block = Block,
		NodeBlock = Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		// ...
	}

	impl frame_system::Config for Runtime {
		..Default
	}
);
```

This could merge well with @kianenigma's idea for declaring default pallet config values...

**Comment by kianenigma** [comment]
> Huh, I just figured that maybe we can even modify the `construct_runtime!` macro to support some custom syntax to configure the test config, something like so:
> 
> ```rust
> construct_runtime!(
> 	pub enum Runtime where
> 		Block = Block,
> 		NodeBlock = Block,
> 		UncheckedExtrinsic = UncheckedExtrinsic
> 	{
> 		// ...
> 	}
> 
> 	impl frame_system::Config for Runtime {
> 		..Default
> 	}
> );
> ```
> 
> This could merge well with @kianenigma's idea for declaring default pallet config values...

Sounds good to me. The syntax might be a bit tricky. `impl frame_system::Config for Runtime {}` would imply that we want to support internals other than `..Default` as well? if no, then I prefer something like this: 

```
construct_runtime!(
	pub enum Runtime where
		Block = Block,
		NodeBlock = Block,
		UncheckedExtrinsic = UncheckedExtrinsic
	{
		System: frame_system{...parts, DefaultImpl},
	}
```

**Comment by thiolliere** [comment]
about:
```
	impl frame_system::Config for Runtime {
		..Default
	}
```
I think it is better to have a rust syntax so we can parse easily.
Maybe just omit the generic type where we want default or write something like `type Foo = PALLET_DEFAULT`

File: frame/support/procedural/src/test_setup.rs (hunk 2 above)
Comment (thiolliere) [reply frame/support/procedural/src/test_setup.rs:151]: unfortunately the macro `pallet` already rely on both frame-system and frame-support.

I was thinking the best would be to actually merge both crates. But I agree improving the error message would be great.

**Pull request closed without merging**